Cover Static abstract interfaces

BillWagner opened the pull request:
Fixes #26735
Fixes #26757
Fixes #24078

- Add an exploration tutorial to introduce static abstract interfaces as a C# 10 preview feature
- Briefly describe the static abstract methods feature in the "What's new" article.

This is the final feature update for C# 10, so this finishes the "what's new" article for that release.

[Internal preview - tutorial](https://review.docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/static-abstract-interface-methods?branch=pr-en-us-27503)
[internal preview - what's new](https://review.docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-10?branch=pr-en-us-27503)

BillWagner committed:
Finish first draft

BillWagner committed:
Proofread and check samples

BillWagner committed:
fix build warnings

BillWagner commented:
ping @adegeo

gfoidl commented on docs/csharp/whats-new/tutorials/snippets/staticinterfaces/Utilities.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    // <MidPoint>
+    public static T MidPoint<T>(T left, T right)
+        where T : INumber<T> => (left + right) / (T.One + T.One);
```
I know this is just an example, but it may suffer from overflow.
Should this be written to avoid overflow*? But it complicates the example, so maybe just
```suggestion
        where T : INumber<T> => (left + right) / (T.One + T.One);  // note: the addition of left, and right may overflow here, it's just for demonstration purposes
```
Hm, I don't like...but the docs should guide people to good and robust code, so somehow this should be mentionend.

Or use a different example than _Midpoint_.

\* for instance: right + (left - right) / 2

gfoidl commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
```diff
@@ -0,0 +1,147 @@
+You’ll need to set up your machine to run .NET 6, including the C# 10 compiler. The C# 10 compiler is available starting with [Visual Studio 2022](https://visualstudio.microsoft.com/downloads/) or the [.NET 6 SDK](https://dotnet.microsoft.com/download).
+
+> [!IMPORTANT]
+> *static abstract members in interfaces* is a preview feature. You must Add the `<EnablePreviewFeatures>True</EnablePreviewFeatures>` in your project file. See [Preview features](https://aka.ms/dotnet-warnings/preview-features)for more information. You can experiment with this feature, and the experimental libraries that use it. We will use feedback from the preview cycles to improve the feature ...
```
Nit: space after link missing
```suggestion
> *static abstract members in interfaces* is a preview feature. You must Add the `<EnablePreviewFeatures>True</EnablePreviewFeatures>` in your project file. See [Preview features](https://aka.ms/dotnet-warnings/preview-features) for more information. You can experiment with this feature, and the experimental libraries that use it. We will use feedback from the preview cycles to improve the feature, including possibly making changes, before its general release.
```

gfoidl commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
````diff
@@ -0,0 +1,147 @@
+Let's start with an example. The following method returns the midpoint of two `double` numbers:
+
+```csharp
+public static double MidPoint(double left, double right) =?
````
```suggestion
public static double MidPoint(double left, double right) =>
```

gfoidl commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
```diff
@@ -0,0 +1,147 @@
+
+Any type that implements the `INumber` interface must include a definition for `operator +`, and `operator /`. The denominator is defined by `(T.One + T.One)` to create the value `2` for any numeric type. The readonly properties `Zero` and `One` are also implemented by every type that implements `INumber`. Using that property rather than `2` forces the denominator to be the same type as the two p ...
+
+You define static abstract members in an interface using familiar syntax: You add the `static` and `abstract` modifiers to any static member that doesn't provide an implementation. The following example defines an `IBoolean<T>` interface that can be applied to any type that overrides `operator true` and `operator false`:
```
ad `IBoolean<T>`: where is the code for this?
The included snippet and the following text is about `IGetNext<T>`. Do I miss something here?

BillWagner replied:
Good catch. I had used a different scenario in my earlier draft, but it wasn't as clear. I'll update the text.

gfoidl commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
```diff
@@ -0,0 +1,147 @@
+
+## Generic math and experimental features
+
+The motivating scenario for allowing static methods, including operators, in interfaces is to support *generic math* algorithms. The [System.Runtime.Experimental](https://www.nuget.org/packages/System.Runtime.Experimental/) NuGet package contains interface definitions for many arithmetic operators, and derived interfaces that combine many arithmetic operators in an `INumber<T>` interface. Let's a ...
```
```suggestion
The motivating scenario for allowing static methods, including operators, in interfaces is to support *generic math* algorithms. The [System.Runtime.Experimental](https://www.nuget.org/packages/System.Runtime.Experimental/) NuGet package contains interface definitions for many arithmetic operators, and derived interfaces that combine many arithmetic operators in an `INumber<T>` interface. Let's apply those types to build a `Point<T>` record that can use any numeric type for `T`. The point can be moved by some `XOffset` and `YOffset` using the `+` operator.
```
?
I'm fine with class too (as in IL it's a class anyway), but why don't type it as _record_ when it's declared that way in the code?

gfoidl commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
````diff
@@ -0,0 +1,147 @@
+}
+```
+
+You use the `record` type for both the `Move<T>` and `Point<T>` types: Both store two values, and they represent data storage rather than sophisticated behavior. The implementation of `operator +` would look like the following code:
````
```suggestion
You use the `record` type for both the `Translation<T>` and `Point<T>` types: Both store two values, and they represent data storage rather than sophisticated behavior. The implementation of `operator +` would look like the following code:
```

gfoidl commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
````diff
@@ -0,0 +1,147 @@
+public record Translation<T>(T XOffset, T YOffset) where T : IAdditionOperators<T, T, T>;
+```
+
+The `IAdditionOperators<T,T,T>` constraint prevents a developer using your class from creating a `Translation` using a type that doesn't meet the constraint for the addition to a point. You've added the necessary constraints to the type parameter for `Translation<T>` and `Point<T>` so this code works. You can test by adding code like the following above the declarations of `Translation` and `Poin ...
````
```suggestion
The `IAdditionOperators<T, T, T>` constraint prevents a developer using your class from creating a `Translation` using a type that doesn't meet the constraint for the addition to a point. You've added the necessary constraints to the type parameter for `Translation<T>` and `Point<T>` so this code works. You can test by adding code like the following above the declarations of `Translation` and `Point` in your *Program.cs* file:
```

gfoidl commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
````diff
@@ -0,0 +1,147 @@
+    new Translation<T>(XOffset: 0, YOffset: 0);
+```
+
+The preceding code won't compile, because `0` depends on the type. The answer: Use `IAdditiveIdentity<T>.AdditiveIdentity` for `0`. That change means that your constraints must now include that `T` implements `IAdditiveIdentity<T>`. Now that you've added that constraint on `Translation<T>`, you need to add the same constraint to `Point<T>`:
````
> The answer: Use `IAdditiveIdentity<T>.AdditiveIdentity` for `0`.

Please insert a code-snippet showing this, to make it very very clear (it's obvious what's meant, but anyway to be complete).

gfoidl commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
```diff
@@ -0,0 +1,147 @@
+
+This sample has given you a look at how the interfaces for generic math compose. You wrote a method that relied on the `INumber<T>` interface so that method could be used with any numeric type. You built a class that relies on the addition interfaces to implement a type that only supports one mathematical operation. That type declares its support for those same interfaces so it can be composed in ...
+
+Experiment with these features and register feedback while they're still in preview. Build generic algorithms that work with any numeric type. Build algorithms using these interfaces where the type argument may only implement a subset of number-like capabilities. You may not build new interfaces that use these capabilities, but you can experiment with using them in your algorithms.
```
> Experiment with these features and register feedback while they're still in preview

Where to "register feedback"?
Should there be a link to https://github.com/dotnet/runtime/issues or what's the intended place to file an issue for this?

BillWagner replied:
Ping @MadsTorgersen @tannergooding @333fred 

Where should I route people to provide feedback?

BillWagner replied:
Got the answer offline.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Günther Foidl <gue@korporal.at>

adegeo commented:
👀

BillWagner committed:
Update docs/csharp/whats-new/tutorials/snippets/staticinterfaces/Utilities.cs

Co-authored-by: Günther Foidl <gue@korporal.at>

BillWagner committed:
Respond to feedback.

Apply suggestions and updates from review.

BillWagner committed:
Add note on potential overflow

adegeo commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
```diff
@@ -0,0 +1,151 @@
+
+Finally, when you're performing addition, it's useful to have a property that defines the additive identity value for that type. There's a new experimental interface for that feature: `IAdditiveIdentity<TSelf, TResult>`. A translation of `{0, 0}` is the additive identity: The resulting point is the same as the left operand. The `IAdditiveIdentity<TSelf, TResult>` interface defines one readonly pr ...
+
+:::code language="csharp" source="./snippets/staticinterfaces/Translation.cs":::
```
For some reason this snippet is showing the `// <ID>` syntax in the preview site. Also, above at line 107 the color coding in github is all messed up. I can't figure out what is breaking the markdown syntax though.

adegeo commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
```diff
@@ -0,0 +1,151 @@
+
+Now that you've added that constraint on `Translation<T>`, you need to add the same constraint to `Point<T>`:
+
+:::code language="csharp" source="./snippets/staticinterfaces/Point.cs":::
```
This one also is showing the `// <ID>` syntax

adegeo commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
```diff
@@ -0,0 +1,151 @@
+
+:::code language="csharp" source="./snippets/staticinterfaces/Point.cs":::
+
+This sample has given you a look at how the interfaces for generic math compose. You wrote a method that relied on the `INumber<T>` interface so that method could be used with any numeric type. You built a class that relies on the addition interfaces to implement a type that only supports one mathematical operation. That type declares its support for those same interfaces so it can be composed in ...
```
nit: Should this be in checkmark syntax? Maybe I'm wrong 😁

```suggestion
This sample has given you a look at how the interfaces for generic math compose. You learned how to:

> [!div class="checklist"]
>
> * Write a method that relied on the `INumber<T>` interface so that method could be used with any numeric type.
> * Build a type that relies on the addition interfaces to implement a type that only supports one mathematical operation.
>   That type declares its support for those same interfaces so it can be composed in other ways. The algorithms are written using the most natural syntax of mathematical operators.
```

adegeo reviewed (approved):
Great tutorial! Don't forget to put it in the TOC! 😄 

I found a few minor things you could improve, and there is a rendering problem of some kind. I'll approve this and let you handle those things.

gfoidl commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
```diff
@@ -0,0 +1,151 @@
+
+:::code language="csharp" source="./snippets/staticinterfaces/Utilities.cs" id="MidPoint":::
+
+Any type that implements the `INumber` interface must include a definition for `operator +`, and `operator /`. The denominator is defined by `(T.One + T.One)` to create the value `2` for any numeric type. The readonly properties `Zero` and `One` are also implemented by every type that implements `INumber`. Using that property rather than `2` forces the denominator to be the same type as the two p ...
```
_Overflow_ happens at the "cpu level". `checked` is to enable "overflow detection" (by looking at the proper state registeres in the cpu) at runtime, and then it woll throw an `OverflowException`.

Try
```c#
int result = int.MaxValue + 1;
```
Then `result == int.MinValue` will be true (note this must happen in `unchecked` as the compiler see the overflow here in static analysis).

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Andy (Steve) De George <67293991+adegeo@users.noreply.github.com>

BillWagner committed:
respond to feedback

BillWagner committed:
add new tutorial to TOC

BillWagner commented:
closing and reopening for a fresh build.

BillWagner committed:
Add link for feedback.

BillWagner committed:
pin to 60.100

BillWagner committed:
pin to 6.0.0

BillWagner committed:
delete global.json

This didn't work in our CI environment.

We don't want to load all the old builds, so wait for the fix on 6.0.101 to merge this.

Youssef1313 commented on docs/csharp/whats-new/csharp-10.md:
```diff
@@ -225,3 +228,12 @@ These types aren't directly represented in metadata. They include annotations th
+## Static abstract members in interfaces
+
+> [!IMPORTANT]
+> *static abstract members in interfaces* is a preview feature. You must Add the `<EnablePreviewFeatures>True</EnablePreviewFeatures>` in your project file. See [Preview features](https://aka.ms/dotnet-warnings/preview-features) for more information. You can experiment with this feature, and the experimental libraries that use it. We will use feedback from the preview cycles to improve the featur ...
```
```suggestion
> *static abstract members in interfaces* is a preview feature. You must add the `<EnablePreviewFeatures>True</EnablePreviewFeatures>` in your project file. See [Preview features](https://aka.ms/dotnet-warnings/preview-features) for more information. You can experiment with this feature, and the experimental libraries that use it. We will use feedback from the preview cycles to improve the feature, including possibly making changes, before its general release.
```

Youssef1313 replied:
> including possibly making changes

Did this mean to say "breaking changes"?

BillWagner replied:
I did mean "making changes". I should just remove that phrase.

Youssef1313 commented on docs/csharp/whats-new/tutorials/snippets/staticinterfaces/staticinterfaces.csproj:
```diff
@@ -0,0 +1,15 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
```
```suggestion
    <OutputType>Exe</OutputType>
    <LangVersion>Preview</LangVersion>
```

BillWagner replied:
Adding the `Preview` element for `LangVersion` isn't necessary. Once the `EnablePreviewFeatures` is set to `true`, the language version tracks to `Preview`.

BillWagner commented:
This is waiting for dotnet/runtime#62840

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Youssef Victor <youssefvictor00@gmail.com>

BillWagner committed:
improve wording.

BillWagner committed:
merge changes

danmoseley commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
````diff
@@ -0,0 +1,163 @@
+Start by creating a new Console application, either by using `dotnet new` or Visual Studio. Next, add the `System.Runtime.Experimental` NuGet package to your project. You'll need to enable preview features to use the types in this package. Enabling preview features also sets the C# language version to "preview", which enables C# 10 preview features. Add the following element to your *csproj* file ...
+
+```xml
+<EnablePreviewFeatures>true</EnablePreviewFeatures>
````
I assume it's obvious that this has to be parented by `<PropertyGroup>` ?

danmoseley commented on docs/csharp/whats-new/tutorials/static-abstract-interface-methods.md:
```diff
@@ -0,0 +1,163 @@
+> * Build a type that relies on the addition interfaces to implement a type that only supports one mathematical operation.
+>   That type declares its support for those same interfaces so it can be composed in other ways. The algorithms are written using the most natural syntax of mathematical operators.
+
+Experiment with these features and register feedback while they're still in preview. You can use the *Send Feedback* menu item in Visual Studio, or create a new [issue](https://github.com/dotnet/roslyn/issues/new/choose) in the roslyn repository on GitHub.  Build generic algorithms that work with any numeric type. Build algorithms using these interfaces where the type argument may only implement  ...
```
I'm not clear whether "you may not" means "it won't work" or "you might not choose to"

BillWagner replied:
Good catch. Reworded.

BillWagner committed:
respond to feedback.

BillWagner committed:
update experimental package.
